api/api.go

```diff
@@ -0,0 +1,200 @@
+package api
+
+import (
+	"errors"
+	"fmt"
+	"github.com/aau-network-security/haaukins/store"
+	"github.com/aau-network-security/haaukins/virtual/vbox"
+	"github.com/dgrijalva/jwt-go"
+	"github.com/gorilla/mux"
+	"net/http"
+	"strconv"
+	"strings"
+)
+
+const (
+	STATUS			= "STATUS"
+	CHALLENGES_KEY	= "CH"
+	token_key		= "test"		//todo its just for testing purpose
+)
+
+var (
+	UnknownTokenErr     = errors.New("Unknown token")
+	ErrorCreateToken	= errors.New("Unable to create the Token")
+)
+
+type LearningMaterialAPI struct {
+	crs 		ClientRequestStore
+	maxRequests	int
+	exStore     store.ExerciseStore
+	vlib  		vbox.Library
+	frontend 	[]store.InstanceConfig
+}
+
+func NewLearningMaterialAPI() (*LearningMaterialAPI, error) {
+	// better approach is to read from a configuration file
+	crs:= NewClientRequestStore()
+	vlib := vbox.NewLibrary("/home/gian/Documents/ova")
+	frontends :=  []store.InstanceConfig{{
+		Image: "kali",
+		MemoryMB: uint(4096),
+	}}
+	ef, err := store.NewExerciseFile("/home/gian/Documents/haaukins_files/configs/exercises.yml")
+	if err != nil {
+		return nil, err
+	}
+	return &LearningMaterialAPI{
+		crs:         *crs,
+		maxRequests: 4,
+		exStore:     ef,
+		vlib:        vlib,
+		frontend:    frontends,
+	}, nil
+}
+
+func (lm *LearningMaterialAPI) Handler() http.Handler {
+	m := mux.NewRouter()
+	m.HandleFunc("/api/{chals}", lm.RequestFromLearningMaterial())
+	return m
+}
+
+func (lm *LearningMaterialAPI) RequestFromLearningMaterial() http.HandlerFunc{
+	var errs error
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		lm.crs.m.RLock()
+		defer lm.crs.m.RUnlock()
+
+		challengesFromLink := mux.Vars(r)["chals"]
+		challenges := strings.Split(challengesFromLink, ",")
+		challengesTag, err := lm.GetChallengesFromRequest(challenges)
+
+		//Bad request (challenge tags don't exist)
+		if err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			w.Header().Set("Content-Type", "text/html; charset=utf-8")
+			w.Write([]byte(errorHTMLTemplate))
+			return
+		}
+
+		//check if the user exists
+		clientR, ok := lm.crs.clientsR[r.Host]
```

Comment:
What if an event is hosted on a high school that has a NAT so that all traffic comes from the same IP address; then only a single environment exists. As I suggested before, use cookies for this.

Also, rate limiting based on IP address suffers from the same problem, so this functionality simply will not work in Haaukins